### ISSUE TITLE
kube-apiserver does not recover after etcd3 -> etcd2 downgrade

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#43366]
This is with k8s 1.6.0-beta.4

I upgraded from etcd -> etcd3 and then back again from etcd3 -> etcd2.  I did this by manually editing the TARGET_STORAGE & TARGET_VERSION from etcd2/2.2.1 -> etcd3/3.0.17 and back again.

I also changed kube-apiserver --storage-backend to etcd3 and then back to etcd2 as I went.

However, kube-apiserver was then unable to perform some operations, such as `kubectl get nodes` or `kubectl get pods -n kube-system`

`kubectl get namespaces` _did_ work (3 namespaces) & `kubectl get pods -n default` _did_ work (no results - it was a new cluster).

Example errors as logged in /etc/kubernetes/kube-apiserver.log:

```
E0320 05:14:25.836138       5 status.go:62] apiserver received an error that is not an metav1.Status: unexpected EOF
I0320 05:14:25.836369       5 wrap.go:75] GET /api/v1/namespaces/kube-system/endpoints/kube-scheduler: (860.182µs) 500
goroutine 2362 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc42367e0e0, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:219 +0xbb
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc42367e0e0, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:198 +0x35
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*responseWriterDelegator).WriteHeader(0xc4237f10e0, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:129 +0x45
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x54f99c0, 0xc420989c80, 0x0, 0x0, 0x32538e0, 0x2, 0x54f28c0, 0xc4215d61f0, 0xc42177bd10, 0x1f4, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:89 +0x162
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x54cb700, 0xc420010100, 0x54f99c0, 0xc420989c80, 0x0, 0x0, 0x32538e0, 0x2, 0x54f28c0, 0xc4215d61f0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:106 +0x103
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(0xc420add540, 0x54cb700, 0xc420010100, 0x54f28c0, 0xc4215d61f0, 0xc42177bd10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:100 +0xd7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.getResourceHandler.func1(0xc4237f1050, 0xc42334c8a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:125 +0x20d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc4237f1050, 0xc42334c8a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:102 +0x1f8
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc4206c66c0, 0x54f2940, 0xc42367e0e0, 0xc42177bd10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:272 +0xba0
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).(k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x54f2940, 0xc42367e0e0, 0xc42177bd10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:120 +0x48
net/http.HandlerFunc.ServeHTTP(0xc4202609a0, 0x54f2940, 0xc42367e0e0, 0xc42177bd10)
	/usr/local/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc420856480, 0x54f2940, 0xc42367e0e0, 0xc42177bd10)
	/usr/local/go/src/net/http/server.go:2022 +0x7f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPanicRecovery.func1(0x54f2940, 0xc42367e0e0, 0xc42177bd10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:74 +0x24a
net/http.HandlerFunc.ServeHTTP(0xc4208575c0, 0x7f4090864a70, 0xc4215d61b8, 0xc42177bd10)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc420842e00, 0x54fb0c0, 0xc4215d61b8, 0xc42177bd10, 0xc42334c780)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:91 +0x8d
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:93 +0x1d1

logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12$\n\x04\n\x00\x12\x00\x12\aFailure\x1a\x0eunexpected EOF\"\x000\xf4\x03\x1a\x00\"\x00"
 [[kube-scheduler/v1.6.0 (linux/amd64) kubernetes/b202120/leader-election] 127.0.0.1:49220]
E0320 05:14:26.243732       5 status.go:62] apiserver received an error that is not an metav1.Status: unexpected EOF
I0320 05:14:26.244061       5 wrap.go:75] GET /api/v1/nodes: (1.675903ms) 500
goroutine 2365 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc42367e850, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:219 +0xbb
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc42367e850, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:198 +0x35
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*responseWriterDelegator).WriteHeader(0xc4237f1a40, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:129 +0x45
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x54f99c0, 0xc420989c80, 0x0, 0x0, 0x32538e0, 0x2, 0x54f28c0, 0xc4215d6250, 0xc42334ad20, 0x1f4, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:89 +0x162
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x54cb700, 0xc420010100, 0x54f99c0, 0xc420989c80, 0x0, 0x0, 0x32538e0, 0x2, 0x54f28c0, 0xc4215d6250, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:106 +0x103
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(0xc420b85b80, 0x54cb700, 0xc420010100, 0x54f28c0, 0xc4215d6250, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:100 +0xd7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1(0xc4237f19b0, 0xc42334d0e0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:336 +0x810
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc4237f19b0, 0xc42334d0e0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:102 +0x1f8
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc4206c66c0, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:272 +0xba0
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).(k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:120 +0x48
net/http.HandlerFunc.ServeHTTP(0xc4202609a0, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc420856480, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:2022 +0x7f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:50 +0x365
net/http.HandlerFunc.ServeHTTP(0xc4207fed80, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:47 +0x1e4d
net/http.HandlerFunc.ServeHTTP(0xc4207fee00, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:73 +0x2c2
net/http.HandlerFunc.ServeHTTP(0xc42083aa50, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.WithRequestContext.func1(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:107 +0xef
net/http.HandlerFunc.ServeHTTP(0xc420842be0, 0x54f2940, 0xc42367e850, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPanicRecovery.func1(0x54f2940, 0xc42367e850, 0xc42334ad20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:74 +0x24a
net/http.HandlerFunc.ServeHTTP(0xc420857500, 0x7f4090864a70, 0xc4215d6208, 0xc42334ad20)
	/usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc420842c80, 0x54fb0c0, 0xc4215d6208, 0xc42334ad20, 0xc42334d080)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:91 +0x8d
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:93 +0x1d1

logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12$\n\x04\n\x00\x12\x00\x12\aFailure\x1a\x0eunexpected EOF\"\x000\xf4\x03\x1a\x00\"\x00"
```

**Comment by liggitt** [comment]
@smarterclayton @wojtek-t is downgrading from etcd3 mode to etcd2 mode supported?

**Comment by liggitt** [comment]
pretty sure running in etcd3 mode with `--storage-media-type` of protobuf (the default in 1.6) puts binary data in etcd, which is not supported by etcd2

**Comment by wojtek-t** [comment]
So downgrade from etcd v3 to etcd v2 is "semi-supported". A bunch of people were advocating that it is not needed (though I don't agree with it).
So we made it supported assuming that:
1. you are still using 'json' as the storage data format (even though the default is protobuf in 1.6)
2. you're not downgrading back to 2.2.1 image, but just using "3.0.17" image and setting ```ETCD_VERSION ``` to 2.2.1 

As there will be stated in the doc, we recommend upgrading to etcd v3 (and then to protobuf storage format) independently from the upgrade to 1.6 release to reduce the risk of those two upgrades.

**Comment by liggitt** [comment]
> we recommend upgrading to etcd v3 (and then to protobuf storage format) independently from the upgrade to 1.6 release to reduce the risk of those two upgrades

the defaults changing to "etcd3" and "protobuf" in 1.6 make it really easy to accidentally upgrade without realizing it, right?

**Comment by wojtek-t** [comment]
That is unfortunately true.

**Comment by liggitt** [comment]
this seems like this is a doc issue (and probably a dupe of https://github.com/kubernetes/kubernetes/issues/41054), albeit a *really* important one

@wojtek-t @mml can we make sure we do all of the following:
* in the release notes for 1.6 (probably at the top, in the largest font possible :), make it clear that action is required related to etcd:
  * upgrade to etcd3 prior to upgrading to 1.6 OR
  * explicitly specify `--storage-type=etcd2 --storage-media-type=application/json` when starting the apiserver
* in the doc for upgrading etcd, make it really clear that you cannot roll back to an etcd2 server (if I'm understanding the comment at https://github.com/kubernetes/kubernetes/issues/43366#issuecomment-287789004 correctly)
* in the doc for upgrading etcd, make it really clear that if you want to be able to roll back from etcd3 to etcd2, you *must* start the 1.6 apiserver with `--storage-media-type=application/json`

**Comment by wojtek-t** [comment]
> in the doc for upgrading etcd, make it really clear that you cannot roll back to an etcd2 server (if I'm understanding the comment at #43366 (comment) correctly)

Technically you can, though it would need to be 2-step rollback:
- first downgrade to "3.0.17" image using version "2.2.1" (assuming you have json)
- after that, you can get back again to "2.2.1" image
Though I don't know if anyone would really want to do this...

**Comment by grodrigues3** [comment]
Do we document the downgrade process somewhere?  Can we get somewhat to update the instructions?

**Comment by mml** [comment]
Most of this is covered in the latest revision of kubernetes/kubernetes.github.io#2767.  I self-assigned and will close once that is merged.

**Comment by mml** [comment]
@liggitt Can you take a look at the doc PR and verify that it's clear enough?

**Comment by ethernetdan** [comment]
@mml @devin-donnelly @wojtek-t can this be done today? this will block the release

**Comment by ethernetdan** [comment]
Doc was merged, #43669 tracks giving a better warning in the script
